### DISCUSSION TITLE
vsgCs -- 3D Tiles and Cesium ion in VSG

### DISCUSSION THREAD

**Discussion by timoore** [vsg-dev/VulkanSceneGraph#690] [General]
A belated Happy New Year! I'm pleased to give the VSG community an early look at [vsgCS](https://github.com/timoore/vsgCs), a library which uses cesium-native to stream assets from Cesium ion and read other 3D Tiles tilesets. It uses some techniques that I think are cool, including:

- building a scene graph node around an external pager which only provides a list of visible tiles
- binding descriptor sets at different levels in the scene graph
- using images in vsgImGui
- loading PNG textures as sRGB, rendering in linear space, and then presenting to an sRGB surface

Lots more to talk about;I hope you find it interesting. I've only tested on Linux, but patches are welcome for Windows and others.

Thanks,
Tim

<img src="https://github.com/timoore/vsgCs/raw/master/doc/img/collioure.png" alt="Collioure, France">

**Comment by robertosfield** [comment]
Very cool.  I haven't had a chance to try it out yet - I will need to download cesium and open an account with them.

The name vsgCS is a bit cryptic, might vsgCesium be more descriptive?  CS might be Compute Science, Compute Shader or Corrupted Snake, but as Cesium isn't written CeSium one has to read the project details to figure out what it might infer.

--

Something for longer term, I've been thinking about how we might easily provide starfields, sun + planets and atmospheric effects in a way that users can go from ground to space and back and get reasonable results.  Rendering of terrain like in vsgCS would need to integrate with this so that the what ever approach taken will need to take this into account.  Essentially I'd like make it easy to get some nice visuals from just loading a ephemeris database and terrain database.

This is a big topic so it's best tackled in a separate thread here, I mention it here as any work that I do in this direction will need to be informed by the needs of projects like vsgCS.

**Comment by timoore** [comment]
vsgCS is cryptic, and is a typo! That's the result of posting at 0100. The actual name of the project is vsgCs, which is still cryptic, but Cs is the chemical symbol for Cesium. I was trying to avoid trademark issues or any hint of affiliation with Cesium Inc.

There are various ephemerides based on Open Scene Graph in the wild: I know of Don Burns' code and that in osgEarth and FlightGear. The latter has a very complete star catalog; the code is GPL, but the catalog comes from the Xephem program which has an MIT license. osgEarth also has a complete atmospheric shader, but it's implemented in terms of osgEarth's virtual program scheme and is hard to tease out into a more conventional form. The question of integrating shaders from a project like vsgCs, whose shaders are not complex but must do more than the default VSG shader, with future environmental effects provided by VSG is indeed going to cause a long discussion. I would suggest looking at what game engines do, but it's not very satisfying.

**Reply by robertosfield** [reply]
I've changed the title of this Discussion thread to reflect the vsgCs naming.  I suspect Cesium would be fine with using the name vsgCesium as it's project the promotes and benefits them, it might be worth asking them.

On the star rendering front I did work in this area so am familiar with the topic, just don't have an open source implementation to share. On the atmosphere front that is still area I need to learn about. Any implementation will need to be MIT compatible.  I'd like to shaders to evolve to be more fine grained and easier to reuse than we have at present.

**Comment by robertosfield** [comment]
@timoore I've been busy with client work so haven't had a chance to follow up on vsgCs yet, I should have  chance later in this week.  

The work I'm doing right now has an overlap with the general areas of rendering paged databases - some of this this work is already partially checked into the VSG and vsgXchange, once I've got far enough I'll start a dedicated thread to discuss it.  There is a component of the work that has a functional overlap with something you've done in vsgCs already - display of logo.

I'm planning to implement a data copyright logo as part of my present work, so it seems this could be something that we want to generally facilitate.  How did you go about rendering the logo for vsgCs?  If such functionality was built into core VSG do you have any thoughts/suggestions about how to go about it/present it?

I'm currently think of a simple logo subgraph that gets assigned to the scene graph after the terrain subgraph and simply renders an unlit textured quad. 

One could possible just have a creaetLogoSubgraph(screenPosition, image) function through to a vsg::Logo node.

**Reply by timoore** [reply]
I actually implemented the logo using vsgImGui. I've been meaning to package up the part that uses VSG to make the Vulkan descriptor set, that ImGui needs to render an image, as a pull request to vsgImGui as it would probably be generally useful. The proper use of CompilerManager::compile() is also a good subject for another thread; I make heavy use of it.

For the ImGui stuff in vsgCS, see [ImageComponent.cpp](https://github.com/timoore/vsgCs/blob/c26c691c9186ff83eac9687c3e49e9b159de912a/src/vsgCs/ImageComponent.cpp)  and [IonIconComponent.cpp](https://github.com/timoore/vsgCs/blob/c26c691c9186ff83eac9687c3e49e9b159de912a/src/CsApp/IonIconComponent.cpp).

**Comment by robertosfield** [comment]
Thanks for the links.  The ImageComponent would be a nice thing to have in vsgImGui so would welcome a PR to vsgImGui.

I have done a quick review and feel that we could probably evolve a few aspects to it we could streamline and generalize a bit more within the context of vsgImGui.  The conversion of RGB to RGBA is something that might not be needed as the VSG can do this when transferring data, but if it is required in this context then perhaps we can move some of the functionality into the core VSG.  These refinements can happen after merging with vsgImGui, or as part of merging a PR.

Do you use vsgImGui for anything other than rendering the logo?

W.r.t vsg::CompileManager/CompileTraversal I think there are opportunities for improvement in various ways, the code is still early in it's life so I expect it to evolve as the community push the envelope in different directions.  With the development of vsg::TransferTask and the performance boost it gave I am also curious about seeing if it could be used in the context of the CompileTraversal.  I can write more about this when the times comes for the dedicated thread.,

**Comment by timoore** [comment]
> Do you use vsgImGui for anything other than rendering the logo?

Not yet. I'm planning on using it shortly to display credits for the data in the scene. In the slightly longer term I'm going to use it for the UI to my viewing program, for choosing viewpoints, lighting options, active layers, etc.

**Reply by robertosfield** [reply]
How do you plan to present the credits?  As a text box, or as text/image just overlaid on screen?  

I ask this as my thoughts on putting up a copyright logo might overlap on these needs so worth rolling into what I write, if it turns out that doing this is really required.

**Comment by robertosfield** [comment]
@timoore did you look at the 3[D Tiles specification](https://cesium.com/why-cesium/3d-tiles/)?

Last year I briefly looked at it but didn't spend time looking what it would entail to implement directly.  From my understanding cesium-ion does this for plus extras.

When I have some time I'll have another look at the 3D Tiles specs to see if it could easily by handled with native VSG functionality.

**Reply by timoore** [reply]
I've read the 3D Tiles spec and I'm also familiar with the osgEarth implementation. The basic .b3dm tile file is basically a wrapper around a binary glTF file that also stores a double precision transformation matrix, links to children, and a variety of other metadata. The newest version of 3D Tiles seems to be doing away with custom file types in favor of vanilla glTF with extensions. If you start with a good glTF reader, then it's probably not too hard to get tiles up on the screen with paging. I don't know if the assimp glTF loader is up to the task or not.

cesium-native implements the glTF and 3D tiles reader / writer and the paging algorithms for managing LODs. It also provides an important extra: the projection of imagery overlays onto the tilesets. The imagery can be streamed from multiple sources, including Bing, TMS, WMS...

**Comment by robertosfield** [comment]
I have finally started diving in getting vsgCs working on my system, first step getting cesium-native compiled.  Oh boy it's a bit of monster, here's the submodules it's pulled down:

```
$ git submodule update --init --recursive
Submodule 'extern/CSPRNG' (https://github.com/Duthomhas/CSPRNG.git) registered for path 'extern/CSPRNG'
Submodule 'extern/Catch2' (https://github.com/catchorg/Catch2.git) registered for path 'extern/Catch2'
Submodule 'extern/GSL' (https://github.com/microsoft/GSL.git) registered for path 'extern/GSL'
Submodule 'extern/KTX-Software' (https://github.com/CesiumGS/KTX-Software.git) registered for path 'extern/KTX-Software'
Submodule 'extern/PicoSHA2' (https://github.com/okdshin/PicoSHA2.git) registered for path 'extern/PicoSHA2'
Submodule 'extern/asyncplusplus' (https://github.com/Amanieu/asyncplusplus.git) registered for path 'extern/asyncplusplus'
Submodule 'extern/cmake-modules' (https://github.com/bilke/cmake-modules.git) registered for path 'extern/cmake-modules'
Submodule 'extern/cpp-httplib' (https://github.com/yhirose/cpp-httplib.git) registered for path 'extern/cpp-httplib'
Submodule 'extern/draco' (https://github.com/google/draco.git) registered for path 'extern/draco'
Submodule 'extern/earcut' (https://github.com/mapbox/earcut.hpp.git) registered for path 'extern/earcut'
Submodule 'extern/expected-lite' (https://github.com/martinmoene/expected-lite.git) registered for path 'extern/expected-lite'
Submodule 'extern/glm' (https://github.com/g-truc/glm.git) registered for path 'extern/glm'
Submodule 'extern/libmorton' (https://github.com/Forceflow/libmorton.git) registered for path 'extern/libmorton'
Submodule 'extern/libwebp' (https://github.com/webmproject/libwebp) registered for path 'extern/libwebp'
Submodule 'extern/rapidjson' (https://github.com/Tencent/rapidjson.git) registered for path 'extern/rapidjson'
Submodule 'extern/s2geometry' (https://github.com/google/s2geometry.git) registered for path 'extern/s2geometry'
Submodule 'extern/spdlog' (https://github.com/gabime/spdlog.git) registered for path 'extern/spdlog'
Submodule 'extern/stb' (https://github.com/nothings/stb.git) registered for path 'extern/stb'
Submodule 'extern/tinyxml2' (https://github.com/leethomason/tinyxml2.git) registered for path 'extern/tinyxml2'
Submodule 'extern/uriparser' (https://github.com/uriparser/uriparser.git) registered for path 'extern/uriparser'
```
Jikes, rather the opposite tack I decided to take for the VSG - trying to take a path that keeps dependencies down but done with care not to end with a massive codebase. 

```
Would these errors be a show stopper?
git-lfs filter-process: 1: git-lfs: not found
fatal: the remote end hung up unexpectedly

&

fatal: Unable to checkout 'b84886590e91f850a32334216322d6a43ba37ca6' in submodule path 'extern/KTX-Software'
```

Down the rabbit hole we go.

**Reply by robertosfield** [reply]
Just run cmake . on cesium-native after pulling in submodules, does this look critical?

```
CMake Error at extern/KTX-Software/CMakeLists.txt:685 (add_subdirectory):
  add_subdirectory given source "tools" which is not an existing directory.


CMake Error at extern/KTX-Software/CMakeLists.txt:689 (add_subdirectory):
  add_subdirectory given source "tests" which is not an existing directory.
```

**Comment by robertosfield** [comment]
So I guess I do need to start debugging the cesium-native build:

```
$ cmake -B build -S .
$ cmake --build build
gmake: Makefile: No such file or directory
gmake: *** No rule to make target 'Makefile'. Stop.
```

Aarrrh,,,

**Comment by timoore** [comment]
It looks like you are having issues with the initial clone and submodule
initialization? I'm doing a clone myself right now, and the submodule stuff
seems to have worked. Otherwise, I didn't have problems building it on
Fedora.

**Reply by robertosfield** [reply]
Still getting the checkout error :-|

**Comment by timoore** [comment]
"Works for me" :) Maybe start over with a fresh clone?

**Reply by robertosfield** [reply]
Just tried again with:

`git clone https://github.com/CesiumGS/cesium-native.git --recurse-submodules`

I get lots of lines suggesting successful cloning of submodules but in amongst them is:

```
Submodule path 'extern/GSL': checked out 'ef0ffefe525a6219ff245d19a832ce06f3fd3504'
git-lfs filter-process: 1: git-lfs: not found
fatal: the remote end hung up unexpectedly
```

And right at the end:

fatal: Unable to checkout 'b84886590e91f850a32334216322d6a43ba37ca6' in submodule path 'extern/KTX-Software'

**Reply by robertosfield** [reply]
Trying again after installing git-lfs...

**Reply by robertosfield** [reply]
Oooo it's checked out all the modules this time, so progress... must have been the git-lfs support that is essential.

**Comment by timoore** [comment]
It looks like KTX-software repo requires git-lfs. I had forgotten that I
installed git-lfs on my Linux machines, and it must be in the default
installation of git for Windows.
Have fun,
Tim

**Comment by robertosfield** [comment]
cesium-ion is now built and installed, and vsgCs has built with this warning:

```
[ 64%] Building CXX object src/vsgCs/CMakeFiles/vsgCs.dir/WorldNode.cpp.o
/home/robert/Dev/vsgCs/src/vsgCs/TilesetNode.cpp: In function ‘vsg::ref_ptr<vsg::Object> {anonymous}::buildTilesetNode(const Value&, vsgCs::JSONObjectFactory*, vsg::ref_ptr<vsg::Object>)’:
/home/robert/Dev/vsgCs/src/vsgCs/TilesetNode.cpp:461:77: warning: ‘this’ pointer is null [-Wnonnull]
  461 |                     vsg::error("expected CSOverly, got ", overlay->className());
      |                                                           ~~~~~~~~~~~~~~~~~~^~
```

**Comment by robertosfield** [comment]
Success!

![Screenshot_20230125_150341](https://user-images.githubusercontent.com/1327952/214598522-b1d1b503-a821-47b8-80fa-fec511197bac.png)

This is what is rendered if I wander over to my little down, our local mountain even has the correct profile.  Houses are a bit flat, but I'll for it that :-)

**Reply by robertosfield** [reply]
Is there a way to set a local file cache so that each load doesn't have to slow pulled down all the tiles?

**Comment by timoore** [comment]
Are there OSM buildings for your town? if so, you can try:
worldviewer --ion-token-file ~/.cesiumiontoken --world-file
vsgCs/tests/osm-buildings.json --no-headlight

Undocumented WIP

**Reply by robertosfield** [reply]
It's almost like I'm there!  

![Screenshot_20230125_182710](https://user-images.githubusercontent.com/1327952/214650802-2dc1dc3c-53b7-4af4-9b57-4fde120d5888.png)

**Comment by timoore** [comment]
Working on it :) I'm going to adapt the example caching accessor in
cesium-native.

**Comment by gwaldron** [comment]
This is great Tim! How do you address the "geometricError" feature? I have been playing with getting screen-space error working in VSG and was about to post a question about it.

**Reply by timoore** [reply]
> This is great Tim! How do you address the "geometricError" feature? I have been playing with getting screen-space error working in VSG and was about to post a question about it.

I don't :sunglasses: All the paging and LOD choices are made by Cesium Native; it does the geometric and screen space error calculations and then, every frame, presents a list of tiles to the application to render.
